src/basic_solver.cpp

```diff
@@ -0,0 +1,166 @@
+#include <pybind11/pybind11.h>
+#include <pybind11/numpy.h>
+#include <pybind11/stl.h>
+#include <pybind11/eigen.h>
+#include <Eigen/Cholesky>
+#include <iostream>
+
+namespace py = pybind11;
+
+
+template<class T, ssize_t N>
+auto _solve_direct_cpp_impl(
+        std::vector<py::array_t<T, py::array::f_style>> bList,
+        std::vector<py::array_t<T, py::array::f_style>> AList,
+        std::vector<py::array_t<T, py::array::f_style>> KList) {
+    // get size info
+    py::buffer_info b_buffer_info = bList[0].request();
+    py::buffer_info A_buffer_info = AList[0].request();
+
+    std::vector<ssize_t> b_shape = b_buffer_info.shape;
+    std::vector<ssize_t> A_shape = A_buffer_info.shape;
+
+    // Allocate and initialize container array for the A transpose A matrix.
+    // For some reason it is faster to allocate a numpy array and then map
+    // it to Eigen than allocating an Eigen array directly
+    // TO DO PF: make sure that A_shape[1] exist. A_shape.at(1)
+    py::array_t<T, py::array::f_style> ATA(std::vector<ssize_t>({A_shape[1]*N, A_shape[1]*N}));
+    py::buffer_info ATA_buffer = ATA.request();
+    T __restrict * ATA_data = static_cast<T *>(ATA_buffer.ptr);
+    // set it all to zero
+    memset(ATA_data, 0, A_shape[1]*N*A_shape[1]*N*sizeof(T));
+    Eigen::Map< Eigen::Matrix<T,
+                              Eigen::Dynamic,
+                              Eigen::Dynamic,
+                              Eigen::ColMajor> > ATA_eig(ATA_data, A_shape[1]*N, A_shape[1]*N);
+
+    // Allocate an array for b, similar to ATA
+    auto ATb = py::array_t<T>(A_shape[1]*N);
+    py::buffer_info ATb_buffer = ATb.request();
+    T __restrict * ATb_data = static_cast<T *>(ATb_buffer.ptr);
+    memset(ATb_data, 0, A_shape[1]*N*sizeof(T));
+
+    // Allocate intermediate arrays that will be reused in the following loops
+    Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic> ATA_small(A_shape[1],A_shape[1]);
+    Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic> K_cross(N, N);
```

Comment:
If you do end up keeping N as a template, then you're missing an opportunity here.  Use `<T, N, N>` rather than Dynamic.